### ISSUE TITLE
Bootstrap on mobile is painfully slow

### ISSUE THREAD

**Issue by flyingmutant** [angular/angular#5198]
There a numerous issues related to bootstrap speed (#2531,  #2618, #4278, #5196), but none of them mentions mobile performance.

While on desktop from my tests bootstrap is kinda-OK-ish (sub-1s first post-«loading…» paint), mobile story is awful right now.

Here are easily reproducible examples for some seed-type (almost empty) apps on OnePlus One. One is a pretty beefy phone: 4-core Snapdragon 801 2.5Ghz, Adreno 330 and 3GB of RAM. Software is Android 5.1 + latest Chrome.

Quick summary is that first real paint happen at ~3 seconds — not a «mobile-first» result, unfortunately.

[angular2-seed](https://github.com/mgechev/angular2-seed) (SystemJS):
<img width="849" alt="2015-11-09 18 12 37" src="https://cloud.githubusercontent.com/assets/73059/11037617/aa98be80-870f-11e5-8b3d-94722c7db9db.png">

---

[angular2-webpack-started](https://github.com/AngularClass/angular2-webpack-starter) (Webpack):
<img width="852" alt="2015-11-09 18 10 49" src="https://cloud.githubusercontent.com/assets/73059/11037636/c1dad10a-870f-11e5-8cd2-21af5bf96cb1.png">

---

P.S. Here is [mithril-quickstart](https://github.com/konsumer/mithril-quickstart) for comparison:
<img width="850" alt="2015-11-09 18 14 12" src="https://cloud.githubusercontent.com/assets/73059/11037670/f01d8ab2-870f-11e5-92a2-09c3bc738d46.png">

**Comment by robwormald** [comment]
@flyingmutant out of curiosity (only because this is the first time I've seen anybody using a OnePlus One in the wild) would you perhaps run the (wholly unrelated to Angular) benchmark mentioned on this (thoroughly depressing) article? https://meta.discourse.org/t/the-state-of-javascript-on-android-in-2015-is-poor/33889 - just as a (wholly unscientific) point of reference. 

I'd like to get a similar suite running for mobile on Angular2, and start tracking across the whole android ecosystem. There will hopefully be a noticeable improvement here with offline template compilation (both in not having to compile templates on-device, and also generally shipping less bytes to be parsed on boot-up).

**Comment by gdi2290** [comment]
> There a numerous issues related to bootstrap speed (#2531, #2618, #4278, #5196), but none of them mentions mobile performance.

I did mention mobile performance in https://github.com/angular/angular/issues/2618#issuecomment-113255603


Minified code, offline template compilation, serviceworker, and server rendering would all help out a lot but the initial file will always be a problem for mobile. There are a few issues talking about tree shaking in Typescript (already a feature in Dart). For [angular2-webpack-starter](https://github.com/AngularClass/angular2-webpack-starter) I also made the mistake of doing dom manipulation before the dom is ready (`bootstrap.ts`) and purposely not providing a production build

**Comment by spock123** [comment]
In my opinion we really cannot say anything about startup-times, until we have a production build setup. The major issue in this case is not the speed of the mobile device, but how many http requests the device has do make before it can bootstrap the application.

As Patrick mentioned, packing everything into 1 or 2 files, compressing stuff, caching it on the device with AppCache or ServiceWorkers will speed up things 10-fold or more.

It's not really going to be an issue and I don't think you'll be able to tell any meaningful difference between different JS frameworks because that's not what matters for 1st time rendering.

**Comment by flyingmutant** [comment]
@spock123 please do not dismiss the issue and do not speculate about random 10-fold improvements.

The test above already *do* pack everything in one file, so the number of requests is minimal. The tests were performed on local network, with zero latency and packet loss, so caching will not change anything (it also does nothing for the so-important first visit). Compressing/minifying also does not help a lot (I've tried that).

Also, the difference in start-up times on mobile is huge for e.g. Angular and Mithril (as you can see from the graphs I've posted). If you'll try to reproduce the results on your mobile device, you'll notice it immediately (waiting several seconds with a blank screen vs. not waiting at all).

@gdi2290 hm, I can't see anything wrong with the [bootstrap.ts](https://github.com/AngularClass/angular2-webpack-starter/blob/master/src/app/bootstrap.ts) – looks quite idiomatic to me. Am I missing something?

@robwormald I've got «mean» of 653ms, with«speed» of 1.53/sec and «error» of ±3674.1%. Looks like until the JIT is warmed up the results are really poor, then things do speed up? Error thing looks crazy (but the results are consistent across reruns).

**Comment by robwormald** [comment]
@flyingmutant okay, that's a good baseline, thanks, and in line with the rest of the commentary on that thread regarding android device performance.

the flame graphs are interesting there - webpack's is obviously smoother, whereas SystemJS has a few "phases", which is strange.

another issue popped up this week where SystemJS bootup was an order of magnitude slower with devtools open (wat?) so investigating that one - i'm curious if it appears faster with devtools closed?

things of note:
- angular2-seed is using the bundles, which are reallllllly super not-optimized for production use, and exist mainly for the purpose of making plunkers easy to use - this explains I think the two-phase thing you're seeing in the timeline, as the entire huge angular bundle loads and reads sync/blocking before the application code kicks off. obviously not ideal.
- same issue with the webpack build - there's a clear delineation between the two sets of source code. This is better during rapid development iteration, as you're not rebuilding the source bundle every time you change an app file, but I suspect that in a production scenario it's going to be better to build a single build, or at least interweave them in a smarter way (think: angular2 core + initial component to render in a single bundle)
- a fair number of the bytes being munged here is stuff that won't ship in a production app (like the compiler) - so that will take a nice chunk off the sheer volume of code currently being parsed at startup.

One interesting project is github.com/rollup/rollup - this is the sort of tooling that will (eventually) give us the best performance in all scenarios - currently we're opting for ease-of-onboarding (bundles) but that focus will shift as we move closer to beta. 

tl:dr; don't panic

**Comment by flyingmutant** [comment]
@robwormald no panic here, just want to bring attention to important thing currently missing from angular's mobile story.

I can not notice any difference with devtools opened/closed — alas, they are opened on my machine while the page is open on the phone.

I have not posted the flamegraph from my own app, but here it is:
<img width="853" alt="2015-11-10 13 14 22" src="https://cloud.githubusercontent.com/assets/73059/11059819/06252804-87ad-11e5-9055-752c611c89a9.png">

All content is bundled together by systemjs-builder; everything is minified and gzipped. Here is the network tab:
<img width="845" alt="2015-11-10 13 16 43" src="https://cloud.githubusercontent.com/assets/73059/11059862/58f9e22c-87ad-11e5-92fe-b895b36b3921.png">

The common thing in all my tests so far is that «Evaluate script» can not start sooner than 700-800 ms (probably due to parsing?). Then the «Evaluate script» itself takes ~1.5 second (lower with webpack, higher with systemjs).

So, to conclude:
- angular must send *much* less code to clients (like, a ton less)
- bundling have to be rethough to have an overhead compatible with old just-concatenate-the-files way (I have high hopes for rollup here as well)

**Comment by robwormald** [comment]
See all those wide 'linkDynamicModule' blocks? That's SystemJS's module resolution churning. Lots of room for optimization there.

**Comment by flyingmutant** [comment]
Sure, but even with Webpack, module loading is not anywhere near 10x faster (1.5x maybe?) — and it needs to be, in order to have sub-1s first paint on mobile.

**Comment by pkozlowski-opensource** [comment]
@flyingmutant @robwormald I think that whatever trick we play (webpack, rollup, removing compiler bits) the crux of the issue is that we are pulling _way_ too much code. We know that ng2 needs to on a diet, simply we didn't focus on it just yet.

**Comment by SynerG** [comment]
Webpack or Rollup are not able to remove unused exports (tree-shaking technique) when applied on the Angular bundles, as the bundles expose ES5 code based on SystemJS or UMD modules. The tools require the import/export statements of ES6 modules to remove unused exports.

@flyingmutant In basic ng2 examples as angular2-seed/angular2-webpack-started I think that the tree-shaking technique should be able to remove a significant amount of unused source code, but you would require to apply it on the ES6 version with proper import/export statements.

Angular exposes the ES6 version in the `/es6/prod` folder of the NPM package.

Maybe you could import that version instead of the bundle and apply Webpack with tree-shake (it requires a little trick to do it properly, check [this article](http://www.2ality.com/2015/12/webpack-tree-shaking.html) for an example).

Rxjs, the biggest Angular dependency, is also developed in TypeScript, so you probably could use its ES6 version to remove unused parts.

I don't know how significant the speed up will be, but it should have an impact. The Angular team will sure improve the boot up process in the future, but this low-hanging fruit could make some difference now.

**Comment by sod** [comment]
The regression is real. Both plunkr forked from angular1 and angular2 landing pages:

http://plnkr.co/edit/55GJb51ZRAxOk7u9LhOI?p=preview angular1-1.5.0 ~65ms
http://plnkr.co/edit/W6l0eZwrNoSri3mjTu32?p=preview angular2-beta.6 ~1100ms

Added to both a `console.time('boostrap')` before first `<script src"...">` and `console.timeEnd('bootstrap')` in first controller/component. So my 4 Core i7 has to chew over a second before angular2 does anything for me.

**Comment by spock123** [comment]
@sod   You're joking, right?

The Angular2 example you're using has no compile step, it's all done in the browser.
This is so that it's easy to get going with playing around with Angular2 code. 

This is NOT how you do it in a production setup.

I suggest you do the same thing but compile your project (for example, use Webpack, Typescript compiler, webpack optimizations), and use Service Workers for caching.

**Comment by sod** [comment]
Removed the compile step

http://plnkr.co/edit/3AJXJMAwE7noe1Mxn5Hb?p=preview angular2-beta.6 ~800ms

**Comment by spock123** [comment]
@sod  you cannot use this a a real-world example.

In your code, you're making 6 http requests before you can even start the bootstrap process.
Also, you're doing in-browser transpiling. You cannot use this setup for anything regarding performance measurements.

I suggest you try doing a build step compilation and bundling. 
Also, put Angular in production mode (default is development mode)

**Comment by sod** [comment]
ok thx for the tipp. Tried https://github.com/AngularClass/angular2-webpack-starter in prod mode and load to bootstrap takes ~200ms. Was not yet able to achive that performance with systemjs + jspm bundle.

**Comment by spock123** [comment]
@sod great stuff!

When the Angular core team gets build time compilation of the templates, you're gonna see even much faster startup times. It's great. 

Also, if you're in production, I would recommend looking into Service workers. This way you can cache all the client code, so that on subsequent visits, your webapp will start up more or less instantly.

Great times ahead :)

**Comment by bnolan** [comment]
We did a bunch of analysis on this (my test device was a Huawei y221, a $30 phone), and found that 80% of the time of loading our app (with ~500 modules loaded to first display) was taken by System.import. This was using production builds. I started on a minimal [system fast-loader](https://github.com/bnolan/system-fast-loader/blob/master/index.js) that doesn't use promises and it made a huge difference on our code base, so we kicked this issue upstream and hopefully system can get a faster loader for production builds.
